### ISSUE TITLE
Support for grouping/pooling samples for reporting

### ISSUE THREAD

**Issue by peterjc** [peterjc/thapbi-pict#290]
Currently our two tabular reports show one row/column for each FASTQ sequenced (replicate) sample. It can be useful to pool (replicate) samples into a single row/column, for example grouping by the original field sample name, or site name, or any other relevant metadata field(s).

For example in https://github.com/peterjc/thapbi-pict/blob/master/examples/recycled_water/metadata.tsv with 384 sequenced samples (each with a row in the metadata table), one might group/pool on site and report source and site:

Source | Site | Sequenced sample count
-- | -- | --
Reservoir | C | 34
Reservoir | G | 32
Reservoir | J | 31
Reservoir | K | 36
Reservoir | L | 29
River | A | 34
River | M | 32
Runoff | B | 37
Runoff | D | 17
Runoff | E | 9
Runoff | F | 11
Runoff | H | 47
Runoff | I | 35

Currently the main values in the table are read counts (per FASTQ sample), so after grouping we could aggregate with the sum (distorts if grouping uneven), mean (might work...), or count of FASTQ non-zero entries (i.e. convert old values to 0 or 1, and sum), or even a boolean present/absent (non-zero or zero).

We could do this within the existing code using Python data structures, indeed I have a prototype on a branch but the initial version does not seem elegant (adds an additional layer of indirection between FASTQ samples and pooling field as dicts).

Or, perhaps shift to pandas (using its built in group by and aggregation support)? This also opens up possibilities for easier graphical output, and options on the tabular reports - We could initially ignore panda's Excel export support and stick fully with the current xlsxwriter based output (for custom coloured Excel), but a hybrid approach may work https://xlsxwriter.readthedocs.io/example_pandas_column_formats.html

**Comment by peterjc** [comment]
Potential API, using ``examples/recycled_water/`` as test case:

```
$ thapbi_pict summary -h
...
  -p COL, --metapooling COL
                        If using metadata, which column values should be used
                        for pooling. Metadata rows with the same value in this
                        column will be pooled as one sample report row. Any
                        other metadata columns requested via -c / --metacols
                        should be consistent. Default 0 is no pooling.
...
```

```
$ thapbi_pict summary -i intermediate/ -o summary/ -m onebp \
-r recycled-water-custom -t metadata.tsv -x 7 -c 1,2
```

vs

```
$ thapbi_pict summary -i intermediate/ -o summary/ -m onebp \
-r recycled-water-pooled -t metadata.tsv -x 7 -c 1,2 -p 2
```

Showing metadata columns 1 and 2 (source and site), pooling on column 2 (site). Goes from 384 rows to 13 rows.

**Comment by peterjc** [comment]
Simpler command line API extension:

```
  -p, --metapooling     If using metadata, should all rows with the same
                        selected column values be pooled? Looks at the -c /
                        --metacols entries except sample filename stems in the
                        -x / --metaindex column. Default no pooling.
```

The ``examples/recycled_water/`` test case becomes:

```
thapbi_pict summary -i intermediate/ -o summary/ -m onebp \
-r recycled-water-pooled -t metadata.tsv -x 7 -c 1,2
```

Current branch gives:

<img width="668" alt="Screenshot 2021-03-19 at 12 13 11" src="https://user-images.githubusercontent.com/63959/111778437-8ca04c00-88ac-11eb-85ee-99d2c6215685.png">

<img width="1086" alt="Screenshot 2021-03-19 at 12 19 08" src="https://user-images.githubusercontent.com/63959/111778928-53b4a700-88ad-11eb-9ab3-34ebf053ee80.png">


Note the sample names values here are cumbersome due to the number of pooled samples, a count might be simpler.

**Comment by peterjc** [comment]
I am leaning towards not doing sample pooling in THAPBI PICT itself, but in a project specific reporting scripts as needed.

**Comment by peterjc** [comment]
Closed by #311 in v0.8.2, see https://github.com/peterjc/thapbi-pict/commits/master/scripts/pooling.py
